Update Drivechains documentation

CryptAxe opened the pull request:
* Updated licenses to BSD 2-Clause license (public domain BIPs are no longer acceptable)
* Typos, spacing, other small updates
* Update data structure examples in documentation to exactly match current codebase
* Describe the new extended serialization transaction type that helps enable BMM
* Remove some sections that seemed to make the documentation more confusing (some things that get very specific are better to see in code and very unlikely to be accurate for any length of time as things are fixed / changed)
* Reorganize some things to make documentation more clear
* Update the ratchet explanation and uncomment it


Mostly updates to bip2-BMM, more work to be done for BIP1-hashrate escrow still.

psztorc commented:
In general I think this is really good.

In the 2nd BIP:

Line 49 the period should be outside the parentheses
Line 66 you changed the number of bytes from 2 to 1. As I understand it that would only allow a sidechain to go back 43 hours which is less than two days. Do you think that is enough time? We could do a crazy trick where there are special mappings for certain values. For example, 254 could map to 1000 blocks ago and 255 could map to 2000 blocks ago. If a sidechain gets into trouble it might have to go forward as before going backward.
On line 70 I liked the sentence "It serves a similar function to Bitcoin's "hashMerkleRoot"." but you've deleted it. The paragraph has a structure of introducing new things and then comparing them to previous things.
Line 78, I think there should be a comma after "sidechain".
Line 101, I don't understand the origin of the 5~7 bytes. It was just three bytes before (1+2).
Line 110, did you do the benchmark to compare 4000 to 8000? 8000 is better if the demands on nodes are sufficiently negligible.
Line 115-123 the table's contents references the 8000 figure. We'd have to change the distances travel to 4000 if we change to 4000.
Old Line 139 (deleted), is it M7s or main:blocks?
Line 137, maybe "affect" instead of "upset"
Line 147-148, I think we changed this a lot. Let us discuss via phone. It seems that we ultimately converged to something much simpler and indeed the funds can just go back to Simon immediately.
I think Line 150 is too unclear and burdens the reader with too much too early. Also there is a typo "may will".
Line 152, you deleted Immediate Expiration, but I thought it was helpful, especially in (my) deleted line 202. I think it is actually the shortest way to explain it.
164, typos "Burton", "profit this"
168, "These tx-fees have a maturity period of 100 side:blocks, the same as regular Bitcoin tx-fees." is probably the clearest.
I think the ratchet stuff is too long. It could contain 158, and the list, and 166 and 168. But that's it I think.
187, I still don't understand why it isnt always 4+1+2 = 7 bytes. And I think the 4 flag bytes should be broken out, to be consistent with everything else. In fact if anything wouldn't the clearest way be "4+35"?
I think your line 196 is a big improvement, but I think the first sentence of my 214 was clear. Instead of "flag" trick we might say "dummy vin", but I think "flag" actually gets the concept across and leads the reader further into the paragraph. And I think it is better to call them "SegWit style transactions" because the concept of SegWit is so large and abstract.
I think the M8_V2 example should stay, but since it has nothing to do with our code it can be moved to m8v2-example.txt or something.

I like the decision to get away from the "bribe" word. Even though I think the bribe word was distinct and helpful.

In general it is different, because a BIP is not really supposed to *explain*; instead it is supposed to specify. But I guess I really can't help myself.

CryptAxe commented:
Fixed everthing you suggested that isn't listed below here https://github.com/drivechain-project/docs/pull/8/commits/c70003a7644e35159ecb921461b8312aaf4cd0c8 and here https://github.com/drivechain-project/docs/pull/8/commits/9a8065939dc486a7ce61bc452ca72f9fe9b39374


> Line 66 you changed the number of bytes from 2 to 1. As I understand it that would only allow a sidechain to go back 43 hours which is less than two days. Do you think that is enough time? We could do a crazy trick where there are special mappings for certain values. For example, 254 could map to 1000 blocks ago and 255 could map to 2000 blocks ago. If a sidechain gets into trouble it might have to go forward as before going backward.

After last time we talked I thought we decided that the max prevBlockRef would be 100. It's very easy to update though, do you have a new idea on many blocks back we should allow a fork to start? The largest reorg I know of was 22 blocks.

> On line 70 I liked the sentence "It serves a similar function to Bitcoin's "hashMerkleRoot"." but you've deleted it. The paragraph has a structure of introducing new things and then comparing them to previous things.

I think it's confusing to compare h* to a hashMerkleRoot. We're indicating the contents of the block, not the root of a merkle tree.


> Line 101, I don't understand the origin of the 5~7 bytes. It was just three bytes before (1+2).

That is part of the spec for the new extended serialization transactions (https://github.com/drivechain-project/bitcoin/blob/mainchainBMM/src/primitives/transaction.h#L180-L229)
The "Mini-Header" includes the h*, prevBlockRef (which will take use more bytes if we change the max prevBlockRef), chainIndex, and bytes to identify the data (which were not included in the old spec). Depending on the values of prevBlockRef and chainIndex, the number of bytes required to represent (and later interpret) the data will change.

>  110, did you do the benchmark to compare 4000 to 8000? 8000 is better if the demands on nodes are sufficiently negligible.

Not yet. What would the main benefit of tracking 8,000 blocks worth of BMM data be compared to 4,000? 

> Old Line 139 (deleted), is it M7s or main:blocks?

I removed that because I added a section that explains how it works see "coinbase cache"

> Line 137, maybe "affect" instead of "upset"

I didn't change that but I'll update it to affect, that's probably a better word.


>  Line 147-148, I think we changed this a lot. Let us discuss via phone. It seems that we ultimately converged to something much simpler and indeed the funds can just go back to Simon immediately.

Yes I removed this because it is explained in detail as a part of the new extended serialization transaction type.


> 168, "These tx-fees have a maturity period of 100 side:blocks, the same as regular Bitcoin tx-fees." is probably the clearest.

In general, critical data transactions require 100 blocks of maturity. For our purpose they also require 100 "blocks atop". They must meet the requirements here: https://github.com/drivechain-project/bitcoin/blob/mainchainBMM/src/consensus/tx_verify.cpp#L228-L233 and here: https://github.com/drivechain-project/bitcoin/blob/1eb389d738bca242d15ffbd80db7872da3034dc0/src/validation.cpp#L602-L606 so I think it's important to point that out, but the wording could probably be improved... hmmm

And this isn't a rule for tx-fees it is a rule for coinbase outputs in the case of Bitcoin, or critical data transaction outputs in our case. I understand what you mean, but I think anyone else reading the documentation would be confused by that wording.


> I think your line 196 is a big improvement, but I think the first sentence of my 214 was clear. Instead of "flag" trick we might say "dummy vin", but I think "flag" actually gets the concept across and leads the reader further into the paragraph. ...

It's not correct to describe it that way. There is a dummy bit where the vin would be, and then there is a flag which indicates the kind of extended serialization data which follows. The flag is a bit pattern used for bitwise operations, and separate from the dummy vin byte.  Segwit's flag is 0001, our flag is 0010.

>  And I think it is better to call them "SegWit style transactions" because the concept of SegWit is so large and abstract.

I'll work on improving this whole section, good point.
